In [2]:
 # Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
path = r'C:\Users\User\02.25 Instacart Basket Analysis'

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [7]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_dow  order_hour_of_day  \
0           0   2539329        1             1          2                  8   
1           1   2398795        1             2          3                  7   
2           2    473747        1             3          3                 12   
3           3   2254736        1             4          4                  7   
4           4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [8]:
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [9]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [10]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [11]:
df_test = pd.DataFrame()

In [12]:
df_test['mix'] = ['a', 'b', 1, True]

In [13]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [14]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [15]:
df_test['mix'] = df_test['mix'].astype('str')

In [16]:
df_test.info

<bound method DataFrame.info of     mix
0     a
1     b
2     1
3  True>

In [17]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [18]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [19]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [20]:
df_prods.shape

(49693, 5)

In [21]:
df_prods.loc[df_prods['product_id'] == 33664, 'prices'] = 7.9

In [22]:
df_prods.loc[df_prods['product_id'] == 21553, 'prices'] = 14.9

In [23]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [24]:
df_prods_clean.shape

(49677, 5)

In [25]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [26]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [27]:
df_prods_clean.shape

(49677, 5)

In [28]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [29]:
df_prods_clean_no_dups.shape

(49672, 5)

# 4.5: Data Consistency Checks Task

# 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further. Tip: Keep an eye on min and max values!

In [32]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [33]:
df_ords.tail()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0

In [34]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

After checking the max and min for every column, there doesn’t appear to be anything noticeably off. The only notable point is that the count for days_since_prior_order is lower than the total number of rows, indicating the presence of null values. However, these null values are actually meaningful, as they signify that zero days have passed since a specific customer placed a new order. These NaN values can be replaced with 0, ensuring that every row in the column contains a valid value.

# 3. Check for mixed-type data in your df_ords dataframe.

In [37]:

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

It doenst appear to be any mix-type columns

# 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

No mixed-type data observed

# 5. Run a check for missing values in your df_ords dataframe.
# In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [42]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Here is possible to observe that there are 206,209 null values in the column days_since_prior_order.

These null values are actually meaningful, as they signify that zero days have passed since a specific customer placed a new order. These NaN values can be replaced with 0, ensuring that every row in the column contains a valid value.

# 6. Address the missing values using an appropriate method.
# In a markdown cell, explain why you used your method of choice.


In [46]:
df_ords_clean = df_ords
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == True
df_ords_clean

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  new_customer  
0                           NaN          True  
1                          15.0         False  
2                          21.0         False  
3                          29.0         False  
4                          28.0         False  
...                         ...           ...  
3421078                    29.0         False  
3421079                    30.0         False  
3421080                    18.0         False  
3421081                     7.0         False  
3421082                    30.0         False  

[3421083 rows x 7 columns]

The Nan values in the days_since_prior_order column represent new customers, so the best aproach to these values is to create a new boolean column named new_customer, which will be True is Nan vlaue is present or false if not.

# 7. Run a check for duplicate values in your df_ords data.
# In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [49]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [50]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

There don’t appear to be any duplicate values in the dataframe.

# 8. Address the duplicates using an appropriate method.
# In a markdown cell, explain why you used your method of choice.

There are no duplicates values!

# 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [55]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [56]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))